src/lib/context.js

```diff
@@ -57,39 +57,47 @@ export class Context {
           () => langs);
   }
 
-  _resolve(langs, id, args, formatter) {
+  _resolve(langs, keys, formatter, resolved = []) {
     const lang = langs[0];
 
     if (!lang) {
-      this._env.emit('notfounderror', new L10nError(
-        '"' + id + '"' + ' not found in any language', id), this);
-      if (formatter === this._formatEntity) {
-        return { value: id, attrs: null };
-      } else {
-        return id;
-      }
+      return reportMissing.call(this, keys, formatter, resolved);
     }
 
-    const entity = this._getEntity(lang, id);
+    let hasUnresolved = false;
+
+    keys.forEach((key, i) => {
+      if (i in resolved) {
```

Comment:
GC costs CPU. I would prefer to avoid changing algorithm that is better optimized for one that requires more from the engine. I don't think it would be cleaner since the resolved array matches the keys array and they zip well together and I'm just iterating over languages (if I need to), but keeping those two by reference.